linkerd/app/inbound/src/policy/store.rs

```diff
@@ -84,53 +75,57 @@ impl Store {
+        // Policies that are dynamically discovered at runtime will expire after
+        // `idle_timeout` to prevent holding policy watches indefinitely for
+        // ports that are generally unused.
+        let cache = {
+            let rxs = ports.into_iter().map(|port| {
+                let watch = watch.clone();
+                let default = default_rx.borrow().clone();
+                let rx =
+                    info_span!("watch", %port).in_scope(|| watch.spawn_with_init(port, default));
                 (port, rx)
-            })
-            .collect::<PortMap<_>>();
-
-        let default = match Self::mk_default(default) {
-            Some((tx, rx)) => {
-                tokio::spawn(async move {
-                    tx.closed().await;
-                });
-                Some(rx)
-            }
-            None => None,
+            });
+            Cache::with_permanent_from_iter(idle_timeout, rxs)
         };
 
         Self {
-            default,
-            ports: Arc::new(rxs),
+            cache,
+            default_rx,
+            discover: Some(watch),
         }
     }
 }
 
-impl CheckPolicy for Store {
-    /// Checks that the destination port is configured to allow traffic.
-    ///
-    /// If the port is not explicitly configured, then the default policy is used. If the default
-    /// policy is `deny`, then a `DeniedUnknownPort` error is returned; otherwise an `AllowPolicy`
-    /// is returned that can be used to check whether the connection is permitted via
-    /// [`AllowPolicy::check_authorized`].
-    fn check_policy(&self, dst: OrigDstAddr) -> Result<AllowPolicy, DeniedUnknownPort> {
-        let server = self
-            .ports
-            .get(&dst.port())
-            .cloned()
-            .map(Ok)
-            .unwrap_or_else(|| match &self.default {
-                Some(rx) => Ok(rx.clone()),
-                None => Err(DeniedUnknownPort(dst.port())),
-            })?;
-
-        Ok(AllowPolicy { dst, server })
+impl<S> GetPolicy for Store<S>
+where
+    S: tonic::client::GrpcService<tonic::body::BoxBody, Error = Error>,
+    S: Clone + Send + Sync + 'static,
+    S::Future: Send,
+    S::ResponseBody:
+        http::HttpBody<Data = tonic::codegen::Bytes, Error = Error> + Default + Send + 'static,
+{
+    fn get_policy(&self, dst: OrigDstAddr) -> AllowPolicy {
+        // Lookup the polcify for the target port in the cache. If it doesn't
+        // already exist, we spawn a watch on the API (if it is configured). If
+        // no discovery API is configured we use the default policy.
+        let server =
+            self.cache
+                .get_or_insert_with(dst.port(), |port| match self.discover.clone() {
+                    Some(disco) => info_span!("watch", %port).in_scope(|| {
+                        disco.spawn_with_init(*port, self.default_rx.borrow().clone())
+                    }),
+                    None => self.default_rx.clone(),
+                });
```

Comment:
i wonder if it was actually better not to hit the cache and just return `Cached::uncached` in the case where the policy is fixed and lookups are disabled. In the fixed policy case we're _just_ cloning the default policy, but the current code here will still insert a clone of the default policy into the map for each port. even though we won't spawn a background task in the fixed case, we still put stuff into the map, so worst case, we have a map with `u16::MAX` u16s and `watch::Receiver` clones.

i [worked it out](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=405c0630fcd70f75a8d8c95617626805) and it looks like the actual data stored in such a map would max out at about 1.7MB, plus however much additional overhead a hashmap takes up.

we could avoid this if we just wrote something like this
```suggestion
        // Lookup the polcify for the target port in the cache. If it doesn't
        // already exist, we spawn a watch on the API (if it is configured). If
        // no discovery API is configured we use the default policy.
        let server = match self.discover {
            Some(ref disco) => self.cache.get_or_insert_with(dst.port(), |port| {
                info_span!("watch", %port).in_scope(|| {
                    disco.spawn_with_init(*port, self.default_rx.borrow().clone())
                })
            },
            None => Cached::uncached(self.default_rx.clone()),
        };
```

but, maybe this is a premature optimization that isn't worth using the `uncached` constructor outside of tests? the maximum memory use of the cache _is_ bounded, but it seems unfortunate to have a giant map where every value is the same thing...

Reply:
We don't really care about the fixed policy case being efficient. I'll document that. We expect proxies to always run with policy discovery, basically. I'm really trying to avoid making `Cached::uncached` a thing if at all possible :)

Reply:
yeah...i guess as long as it's not a configuration that will use an *unbounded* amount of memory, it doesn't really matter...